### PR TITLE
feat: allow `array_slice` to take an optional stride parameter

### LINKED ISSUE #10424: `stride` is not optional for new `array_slice` UDF
### Describe the bug

The `array_slice` UDF takes 4 parameters.

https://github.com/apache/datafusion/blob/96487ea0cbb7901a1e4aa18fdf6deb8961319fea/datafusion/functions-array/src/extract.rs#L55-L61

Which means that `args.len()` is always 4 in `array_slice_inner`, even when called with `stride = Expr::Null` or `stride = Expr::Int64(None)`

https://github.com/apache/datafusion/blob/96487ea0cbb7901a1e4aa18fdf6deb8961319fea/datafusion/functions-array/src/extract.rs#L289-L293

### To Reproduce

I encountered this while creating the python wrapper for `datafusion-python` and had to set `stride=1` else the function would panic.

https://github.com/apache/datafusion-python/blob/7ad526caf140f7eb76ec541c903027d49693b58f/src/functions.rs#L104-L111

### Expected behavior

The `stride` behavior in the UDF should match what's available in the CLI

```
CREATE TEMPORARY VIEW data3 AS VALUES ([1.0, 2.0, 3.0, 3.0]), ([4.0, 5.0, 3.0]), ([6.0]);
❯ select * from data3;
+----------------------+
| column1              |
+----------------------+
| [1.0, 2.0, 3.0, 3.0] |
| [4.0, 5.0, 3.0]      |
| [6.0]                |
+----------------------+

> select * from array_slice(data3.column1, -1, 2)
+-----------------------------------------------+
| array_slice(data3.column1,Int64(-1),Int64(2)) |
+-----------------------------------------------+
| []                                            |
| []                                            |
| [6.0]                                         |
+-----------------------------------------------+
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jonahgao**
## Which issue does this PR close?

Closes #10424.

## Rationale for this change
Based on the draft PR #10450, and reuse the existing `make_udf_function` macro as much as possible.

## What changes are included in this PR?

- Allow array_slice to take an optional stride parameter

## Are these changes tested?

Not sure whether it's necessary to write a test for this.

## Are there any user-facing changes?
Yes

### CODE REVIEW

File: datafusion/functions-array/src/macros.rs [hunk 1]
```diff
@@ -106,4 +105,26 @@ macro_rules! make_udf_function {
             }
         }
     };
+    // This pattern does not generate the "fluent expr_fn" style function, it should be provided externally.
+    ($UDF:ty, $SCALAR_UDF_FN:ident) => {
```
Comment (jayzhan211) [inline datafusion/functions-array/src/macros.rs:109]: I would recommend to pull out `create_function` like https://github.com/apache/datafusion/blob/main/datafusion/functions-aggregate/src/macros.rs.

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs [hunk 2]
```diff
@@ -581,7 +581,7 @@ async fn roundtrip_expr_api() -> Result<()> {
             make_array(vec![lit(1), lit(2), lit(3)]),
             lit(1),
             lit(2),
-            lit(1),
+            Some(lit(1)),
```
Comment (jayzhan211) [inline datafusion/proto/tests/cases/roundtrip_logical_plan.rs:584]: Maybe add the example that has None for stride?

File: datafusion/functions-array/src/macros.rs (hunk 1 above)
Comment (jonahgao) [reply datafusion/functions-array/src/macros.rs:84]: Applied. This eliminates some duplication of macro codes. Thank you @jayzhan211 .

File: datafusion/proto/tests/cases/roundtrip_logical_plan.rs (hunk 2 above)
Comment (jonahgao) [reply datafusion/proto/tests/cases/roundtrip_logical_plan.rs:584]: Added.

**Review by jayzhan211 — APPROVED** [review]
👍

**Comment by Michael-J-Ward** [comment]
@jayzhan211, Should I follow this as a template for making UDF arguments optional?

**Review by alamb — APPROVED** [review]
LGTM -- thank you @jonahgao and @jayzhan211

**Pull request merged**